flink-rpc/flink-rpc-akka/src/main/java/org/apache/flink/runtime/rpc/akka/RobustActorSystem.java

```diff
@@ -0,0 +1,106 @@
+import org.apache.flink.util.FatalExitExceptionHandler;
+
+import akka.actor.ActorSystemImpl;
+import akka.actor.BootstrapSetup;
+import akka.actor.setup.ActorSystemSetup;
+import com.typesafe.config.Config;
+import com.typesafe.config.ConfigFactory;
+
+import java.util.Optional;
+
+import scala.Option;
+import scala.concurrent.ExecutionContext;
+
+/**
+ * {@link ActorSystemImpl}} which has a configurable {@link
+ * java.lang.Thread.UncaughtExceptionHandler}.
+ *
+ * <p>The class is abstract because instances need to override {@link
+ * ActorSystemImpl#uncaughtExceptionHandler()}, as this method is called from the super constructor.
+ */
+public abstract class RobustActorSystem extends ActorSystemImpl {
+
+    public RobustActorSystem(
+            String name,
+            Config applicationConfig,
+            ClassLoader classLoader,
+            Option<ExecutionContext> defaultExecutionContext,
+            ActorSystemSetup setup) {
+        super(name, applicationConfig, classLoader, defaultExecutionContext, Option.empty(), setup);
+    }
+
+    public static RobustActorSystem create(String name, Config applicationConfig) {
+        return create(name, applicationConfig, FatalExitExceptionHandler.INSTANCE);
+    }
+
+    @VisibleForTesting
+    static RobustActorSystem create(
+            String name,
+            Config applicationConfig,
+            Thread.UncaughtExceptionHandler uncaughtExceptionHandler) {
+        return create(
+                name,
+                ActorSystemSetup.create(
+                        BootstrapSetup.create(
+                                Optional.empty(),
+                                Optional.of(applicationConfig),
+                                Optional.empty())),
+                Option.apply(uncaughtExceptionHandler));
+    }
+
+    private static RobustActorSystem create(
+            String name,
+            ActorSystemSetup setup,
+            Option<Thread.UncaughtExceptionHandler> uncaughtExceptionHandler) {
+        final Optional<BootstrapSetup> bootstrapSettings = setup.get(BootstrapSetup.class);
+        final ClassLoader classLoader = RobustActorSystem.class.getClassLoader();
+        final Config appConfig =
+                bootstrapSettings
+                        .map(BootstrapSetup::config)
+                        .flatMap(RobustActorSystem::toJavaOptional)
+                        .orElseGet(() -> ConfigFactory.load(classLoader));
+        final Option<ExecutionContext> defaultEC =
+                toScalaOption(
+                        bootstrapSettings
+                                .map(BootstrapSetup::defaultExecutionContext)
+                                .flatMap(RobustActorSystem::toJavaOptional));
+
+        final RobustActorSystem robustActorSystem =
+                new RobustActorSystem(name, appConfig, classLoader, defaultEC, setup) {
+                    @Override
+                    public Thread.UncaughtExceptionHandler uncaughtExceptionHandler() {
```

Comment:
As noted in the javadocs this method is called from the `ActorSystemImpl` constructor, so we can't just pass the handler into the `RobustActorSystem` and have it return it.